File: src/utils/format.ts
import { COMISION_PCT } from '../data/constants';
import { HistorialItem, Usuario } from '../data/types';

export const fmtTime = (ts: string) => {
  const d = new Date(ts);
  const now = new Date();
  const diff = (now.getTime() - d.getTime()) / 60000;
  if (diff < 1) return 'ahora';
  if (diff < 60) return `${Math.floor(diff)}m`;
  if (diff < 1440) return `${Math.floor(diff / 60)}h`;
  return d.toLocaleDateString('es-HN', { day: '2-digit', month: 'short' });
};

export const fmtMoneda = (n: number | null | undefined) => (n == null ? '' : '$' + Number(n).toLocaleString());

export const uid = () => Date.now() + Math.random();

export const rand = (a: number, b: number) => Math.floor(Math.random() * (b - a + 1)) + a;

export const calcComision = (monto: number) => {
  const comision = Number(((monto * COMISION_PCT) / 100).toFixed(2));
  return { comision, montoTransportista: Number((monto - comision).toFixed(2)) };
};

export const avgRating = (historial: HistorialItem[], usuario: Pick<Usuario, 'id' | 'tipo'>): number | null => {
  const campo = usuario.tipo === 'cliente' ? 'calCliente' : 'calTransportista';
  const idCampo = usuario.tipo === 'cliente' ? 'clienteId' : 'transportistaId';
  const vals = historial
    .filter((h) => (h as any)[idCampo] === usuario.id && (h as any)[campo])
    .map((h) => (h as any)[campo].estrellas as number);
  if (!vals.length) return null;
  return vals.reduce((a, b) => a + b, 0) / vals.length;
};

File: lib/utils/format.dart
import 'package:intl/intl.dart';
import '../models/models.dart';

const _mesesEs = ['ene', 'feb', 'mar', 'abr', 'may', 'jun', 'jul', 'ago', 'sep', 'oct', 'nov', 'dic'];

String fmtTime(DateTime ts) {
  final now = DateTime.now();
  final diffMin = now.difference(ts).inMinutes;
  if (diffMin < 1) return 'ahora';
  if (diffMin < 60) return '${diffMin}m';
  if (diffMin < 1440) return '${(diffMin / 60).floor()}h';
  return '${ts.day.toString().padLeft(2, '0')} ${_mesesEs[ts.month - 1]}';
}

/// Fecha y hora completas, para constancias que hay que poder citar después.
String fechaLarga(DateTime d) =>
    '${d.day} ${_mesesEs[d.month - 1]} ${d.year}, ${d.hour.toString().padLeft(2, '0')}:${d.minute.toString().padLeft(2, '0')}';

final _moneyFmt = NumberFormat('#,##0.##', 'en_US');

String fmtMoneda(num? n) => n == null ? '' : '\$${_moneyFmt.format(n)}';

double? avgRating(List<HistorialItem> historial, {required String usuarioId, required TipoUsuario tipo}) {
  final vals = <int>[];
  for (final h in historial) {
    if (tipo == TipoUsuario.cliente && h.clienteId == usuarioId && h.calCliente != null) {
      vals.add(h.calCliente!.estrellas);
    } else if (tipo != TipoUsuario.cliente && h.transportistaId == usuarioId && h.calTransportista != null) {
      vals.add(h.calTransportista!.estrellas);
    }
  }
  if (vals.isEmpty) return null;
  return vals.reduce((a, b) => a + b) / vals.length;
}
